text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,384 @@
+};
+```
+
+```
+[Exposed=Window]
+interface Node : EventTarget {
+  [CEReactions] attribute DOMString? nodeValue;
+  [CEReactions] attribute DOMString? textContent;
+
+  [CEReactions] Node appendChild(Node node);
+  [CEReactions] Node removeChild(Node child);
+
+  ... other attributes and methods ommitted ...
+};
+```
+
+Based upon that WebIDL spec, the WebIDL generator will generate this Rust code:
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, js_name = addEventListener)]
+    fn add_event_listener(this: &EventTarget, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = removeEventListener)]
+    fn remove_event_listener(this: &EventTarget, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = dispatchEvent)]
+    fn dispatch_event(this: &EventTarget, event: Event);
+}
+
+pub trait IEventTarget: AsRef<EventTarget> {
+    #[inline]
+    fn add_event_listener(&self, type: &str, callback: &Closure<FnMut(Event)>, options: bool) {
+        EventTarget::add_event_listener(self.as_ref(), type, callback, options)
+    }
+
+    #[inline]
+    fn remove_event_listener(&self, type: &str, callback: &Closure<FnMut(Event)>, options: bool) {
+        EventTarget::remove_event_listener(self.as_ref(), type, callback, options)
+    }
+
+    #[inline]
+    fn dispatch_event(&self, event: Event) {
+        EventTarget::dispatch_event(self.as_ref(), event)
+    }
+}
+
+impl IEventTarget for EventTarget {}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, getter = nodeName)]
+    fn node_value(this: &Node) -> JsString;
+
+    #[wasm_bindgen(method, getter = textContent)]
+    fn text_content(this: &Node) -> JsString;
+
+    #[wasm_bindgen(method, js_name = appendChild)]
+    fn append_child(this: &Node, node: Node) -> Node;
+
+    #[wasm_bindgen(method, js_name = removeChild)]
+    fn remove_child(this: &Node, child: Node) -> Node;
+}
+
+pub trait INode: IEventTarget + AsRef<Node> {
+    #[inline]
+    fn node_value(&self) -> JsString {
+        Node::node_value(self.as_ref())
+    }
+
+    #[inline]
+    fn text_content(&self) -> JsString {
+        Node::text_content(self.as_ref())
+    }
+
+    #[inline]
+    fn append_child(&self, node: Node) -> Node {
+        Node::append_child(self.as_ref(), node)
+    }
+
+    #[inline]
+    fn remove_child(&self, child: Node) -> Node {
+        Node::remove_child(self.as_ref(), child)
+    }
+}
+
+impl IEventTarget for Node {}
+impl INode for Node {}
+```
+
+Essentially, it does this:
+
+1. It adds concrete private methods to the types (e.g. `EventTarget` and `Node`)
+
+2. It creates a new trait which has the same name as the type, but prefixed with `I` (e.g. `IEventTarget` and `INode`)
+
+3. This trait has an `AsRef<Type>` constraint
+
+4. If the WebIDL interface extends from another interface, then that is also added as a constraint (e.g. `INode` inherits from `IEventTarget`)
```

Comment:
Maybe make clear that you only put direct inheritance as a constraint? For example `IElement` would not have the direct constraint of `IEventTarget + INode`, but only `INode` (the `IEventTarget` constraint would be implied).